packages/vite/src/node/ssr/fetchModule.ts

```diff
@@ -36,33 +37,35 @@ export async function fetchModule(
   }
 
   if (url[0] !== '.' && url[0] !== '/') {
-    const {
-      isProduction,
-      resolve: { dedupe, preserveSymlinks },
-      root,
-      ssr,
-    } = server.config
-    const overrideConditions = ssr.resolve?.externalConditions || []
-
-    const resolveOptions: InternalResolveOptionsWithOverrideConditions = {
-      mainFields: ['main'],
-      conditions: [],
-      overrideConditions: [...overrideConditions, 'production', 'development'],
-      extensions: ['.js', '.cjs', '.json'],
-      dedupe,
-      preserveSymlinks,
-      isBuild: false,
-      isProduction,
-      root,
-      ssrConfig: ssr,
-      packageCache: server.config.packageCache,
-    }
+    const { isProduction, root } = environment.config
+    const { externalConditions, dedupe, preserveSymlinks } =
+      environment.options.resolve
 
     const resolved = tryNodeResolve(
       url,
       importer,
-      { ...resolveOptions, tryEsmOnly: true },
-      false,
+      {
```

Comment:
are these options correct?

this `fetchModule` is what the `DevEnvironment` class uses right?

these options seem to be specifically targeting node? (given the `.cjs` extension and the `main` field usage)

should this be make more generic, or actually could this be made more configurable?
(so that `DevEnvironment` instances could tweak this options as needed?)

Reply:
Or actually, shouldn't this inherit from the environment's options, some fields are indeed inherited:
https://github.com/vitejs/vite/pull/16471/files/2b3329d4aa0b91a5920527a9712b82ae59becba3#diff-fde6fc19368b69a4726cb38039db6dee6abf54bc69e1484c4f6e6c9d952acd9dR40-R42

Reply:
These are the same options that were used for `ssrLoadModule` before. It might be a good idea to provide a way to customize it, but you can also intercept the `fetchModule` call at any time

Reply:
what do you mean by intercepting the call? 🙏 

Right now what I am doing is creating a new `DevEnvironment` instance ([source](https://github.com/dario-piotrowicz/vite-environment-6.0.0-alpha-experimentations/blob/66a383dcaa5d5ba5e8b6c57669c7907cb324493a/packages/vite-environment-provider-workerd/src/index.ts#L153-L155)):
```ts
const devEnv = new ViteDevEnvironment(name, config, {
  hot,
}) as DevEnvironment;
```

and then using the instance's `fetchModule` in order to let vite apply its module resolution ([source](https://github.com/dario-piotrowicz/vite-environment-6.0.0-alpha-experimentations/blob/66a383dcaa5d5ba5e8b6c57669c7907cb324493a/packages/vite-environment-provider-workerd/src/index.ts#L124)):
```ts
const result: any = await devEnv.fetchModule(...(args as [any, any]));
```

So that I can feed this result back to the module runner ([source](https://github.com/dario-piotrowicz/vite-environment-6.0.0-alpha-experimentations/blob/66a383dcaa5d5ba5e8b6c57669c7907cb324493a/packages/vite-environment-provider-workerd/src/worker/index.ts#L58-L69)):
```ts
transport: {
  fetchModule: async (...args) => {
    const response = await env.__viteFetchModule.fetch(
      new Request('http://localhost', {
        method: 'POST',
        body: JSON.stringify(args),
      }),
    );
    const result = response.json();
    return result as any;
  },
},
```

is there a different way which would allow me to fetch the modules with my own resolve options?

Reply:
@sheremet-va me and Igor have been looking more into this and we agree that the default node environment should preserve the `ssrLoadModule` behavior.

The `fetchModule` implementation in this class is however generally useful for custom environments, and in fact it's already exposed as part of the Environment API, so custom environments get to benefit from it.

For the custom environments to however truly benefit from this implementation, the environment config needs to be honored as I've pointed out above.

In order to preserve the behavior for the default node environment, we just need to ensure that it's default node environment is configured in a way that preserves the original `ssrLoadModule` implementation.

Would this work for you or are we missing anything else that complicates the matter? Thank you

Reply:
Yes, I agree that it needs to be customizable.

Reply:
I guess here we can only expose these:

- `mainFields` -> `externalMainFields`
- `extensions` -> `externalExtensions`
- `tryEsmOnly` -> `externalTryEsmOnly`

All other fields are related to externalization and not to the environment. Is this ok for you?

Reply:
External modules are modules that are consumed by the runtime that runs the code in dev. During build, these modules will be kept as is instead of bundling them:

```ts
// bundled module - "inlined" in dev
import './inlined.js'
// external module
import 'vue'
// external module
import 'vue-router'
```

The Vite module runner emulates the same module to keep the code working the same way as it does in build. If we don't have `external*` fields, then Vite module resolution might not follow the host module resolution and in the example above, the `vue` from inlined module and `vue` from the external module might resolve to different files which will not happen when running built code because _all imports_ are resolved by the runtime in that case (for example, Node.js). If the expected runtime doesn't support importing external modules, it should not support them in the Vite module runner with `noExternal` for example.

> What should be the value of these external* options? And who determines that? Are the values just a copy of the original mainFields and friends?

The runtime that consumes the built code should define them during dev for the module runner (which is a dev-only feature).

Reply:
After the conversation we had on discord ([thread](https://discord.com/channels/804011606160703521/831456449632534538/threads/1247574406637817917)) I believe that we've been stumbled across this method and used it in the wrong context, assuming that it was a general purpose / external API ok to be used in order to reuse Vite's module resolution.

My understanding now is that this is a nodejs specific implementation used by Vite itself that just so happens to be accessible externally and that we've been trying to repurpose for the wrong reasons.

Vite does not provide an API nor does make any assumptions on how external modules should be resolved and that it is on us to define and implement some external module resolution based on the specific logic and requirements of our runtime.

As @sheremet-va suggested maybe via the `import-meta-resolve` package or by creating a resolver via `createResolver` (assuming that that is available on the DevEnvironment).

So I believe that the current implementation of `tryNodeResolve` is valid as is and no additional customization capabilities should actually be added.

I believe the confusion here was caused by a misunderstanding of the Vite API on my part, @sheremet-va sorry for the headache and the confusion, thanks a lot for the clarifications and patience on this 🙏 

I think it's ok to resolve this conversation, keep `tryNodeResolve` as is and we'll resolve things on our end.

I'd just wait to make sure @IgorMinar's also ok with this outcome 🙂

Reply:
`createIdResolver` is now exposed in vite@6.0.0-alpha.18 in case you need to use it. I'm not sure we'll want to expose this to be honest. It is useful in some of our internal plugins and `config.createResolver` has external usage but it was always hard to support. But if you end up needing it, we'll discuss to include it. We'll probably need a replacement for `config.createResolver` anyways.

Reply:
thanks so much @patak-dev 🙂👍 

I'll have a look and let you know, it's likely to be a good solution for our issue, but we might also not really need it, I'll explore the possible alternatives and let you know 🙂 

https://github.com/dario-piotrowicz/vite-environment-6.0.0-alpha-experimentations/issues/5

(PS: as I mentioned on discord we're currently focusing on a different project for a bit, so I don't have a lot of time to dedicate to this in the next few days, but if you need us to tell you soon if we do need this `createResolver` API I will prioritize and make sure I work on this ASAP, please just let me know)

Reply:
There is no rush, good luck with the other project!

Reply:
You can see what `createIdResolver` does here: https://github.com/vitejs/vite/blob/bbca0b402959ceeb71e08588808fab045f58b887/packages/vite/src/node/idResolver.ts#L33. It is properly using the environment configuration (things like `resolve.conditions`, `resolve.alias` still needs to be worked out but that will also work). It is the alias+resolve plugins, instead of the whole plugin pipeline for the environment so custom resolving added by users won't apply (optimized deps ids also don't play a role here). It is currently implemented using these two plugins to reuse the implementation, but I think it would be better at some point to have the resolve logic as free functions we can call instead of needing a plugins container for two plugins.
Internally, we are using this resolver to implement custom resolution. See for example resolving of CSS preprocessor files, that have their own rules: https://github.com/vitejs/vite/blob/bbca0b402959ceeb71e08588808fab045f58b887/packages/vite/src/node/plugins/css.ts#L1067. We don't let users hook into these.

We also use it to make esbuild follow Vite alias and resolve rules during dependency optimization, so it matches what Rollup will do during build. See here https://github.com/vitejs/vite/blob/bbca0b402959ceeb71e08588808fab045f58b887/packages/vite/src/node/optimizer/esbuildDepPlugin.ts#L68. I think checking this file should be useful for you. We are also dealing with externals, CJS and esm, etc.

Reply:
Perfect! I'll need to thoroughly look into those! thanks so much 🙏

Reply:
PS: I am not sure why but I can't resolve this thread, if you want please feel free to resolve this thread since the original issue has been resolved and this might just add unnecessary noise to this PR at this point (and I can ping you on discord for any further discussion on the resolver situation 🙂)

Reply:
> Anyways he made a great point in the fact that we really need to have the module resolution applied here be the exact same as the one that will be used for the production build (so that this behaves as close as possible to what gets actually deployed), so:

But Vite leaves external modules as is (so, `vue` will be left as `vue`, not `/node_modules/vue`) in production, there is no Vite resolution during build for SSR 🤔 

The workerd does the resolution. You said that it works in production somehow

Reply:
> The workerd does the resolution. You said that it works in production somehow.

@sheremet-va not really 😅 I said that workerd doesn't do the resolution and that in production the modules are bundled in and not external. The issue is that they need to be external during development because they are cjs (which the Vite dev server can't handle).

So during build they are resolved by the Vite resolution and during development they unfortunately need to be external, but they should be resolved to, as close as possible, the same modules that would be used during the build time resolution